internal/ethapi/api.go

```diff
@@ -1049,6 +1084,80 @@ func (s *BlockChainAPI) Call(ctx context.Context, args TransactionArgs, blockNrO
 	return result.Return(), result.Err
 }
 
+// BatchCallConfig is the config object to be passed to eth_batchCall.
+type BatchCallConfig struct {
+	Block          rpc.BlockNumberOrHash
+	StateOverrides *StateOverride
+	Calls          []BatchCallArgs
+}
+
+// BatchCallArgs is the object specifying each call within eth_batchCall. It
+// extends TransactionArgs with the list of block metadata overrides.
+type BatchCallArgs struct {
+	TransactionArgs
+	BlockOverrides *BlockOverrides
+}
+
+// CallResult is the result of one call.
+type CallResult struct {
+	Return hexutil.Bytes
+	Error  error
+}
+
+// BatchCall executes a series of transactions on the state of a given block as base.
+// The base state can be overridden once before transactions are executed.
+//
+// Additionally, each call can override block context fields such as number.
+//
+// Note, this function doesn't make any changes in the state/blockchain and is
+// useful to execute and retrieve values.
+func (s *BlockChainAPI) BatchCall(ctx context.Context, config BatchCallConfig) ([]CallResult, error) {
+	state, header, err := s.b.StateAndHeaderByNumberOrHash(ctx, config.Block)
+	if state == nil || err != nil {
+		return nil, err
+	}
+	// State overrides are applied once before all calls
+	if err := config.StateOverrides.Apply(state); err != nil {
+		return nil, err
+	}
+	// Setup context so it may be cancelled before the calls completed
+	// or, in case of unmetered gas, setup a context with a timeout.
+	var (
+		cancel  context.CancelFunc
+		timeout = s.b.RPCEVMTimeout()
+	)
+	if timeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, timeout)
+	} else {
+		ctx, cancel = context.WithCancel(ctx)
+	}
+	// Make sure the context is cancelled when the call has completed
+	// this makes sure resources are cleaned up.
+	defer cancel()
+	var (
+		results []CallResult
+		// Each tx and all the series of txes shouldn't consume more gas than cap
+		globalGasCap = s.b.RPCGasCap()
+		gp           = new(core.GasPool).AddGas(globalGasCap)
+	)
+	for _, call := range config.Calls {
+		blockContext := core.NewEVMBlockContext(header, NewChainContext(ctx, s.b), nil)
+		if call.BlockOverrides != nil {
```

Comment:
I don't quite understand why we have call-level block overrides. In practice these calls usually have the same block context if they want to be put in a single block by intention?

Reply:
Because users can simulate the case that transactions are included in different blocks? If so I think this design makes sense.

Reply:
Micah also asked a similar question here: https://github.com/ethereum/execution-apis/pull/312#discussion_r981964051

Reply:
On one hand, it makes sense. For example, if you want to experiment with a time-locked contract. First you create it, then two years pass, now you want to interact again. 
It opens up a lot of potential uses which does not fit inside a single block. 

However, it might also be a footgun. If you want to simulate a sequence where

1. Block `n`: A contract `X` is selfdestructed, 
2. Block `n+1`, contrat `X` is resurrected. 

The two steps can never happen in _one_ block. The question is: what happens in the batch-call? Is it possible to make the two calls execute correctly, or will it be some form of "time-shifted single block", where you can override the time and number, but state-processing-wise it's still the same block.... ?

Reply:
I also am concerned that not all clients will have an easy time implementing this as currently specified.  I suspect all clients could have two distinct blocks that they execute against some existing block's post-state, but not all clients may be able to simulate a series of transcations against a cohesive state when the transaction's don't share block fields.

Would be great to get other client feedback on this to verify, but without any feedback I would assume the worst that this will be "hard" to implement in some clients.  Having writing some Nethermind plugins, my gut suggests that this would be challenging to do with Nethermind for example.

Reply:
Revisiting this. Actually, I was confused when I wrote that. So the destruct + resurrect _can_ happen within one block, just not within one transaction. 

The point remains, though. There are subtle differences whether a two executions are same block or different blocks -- such as whether the block coinbase has been rewarded with the fees.

Reply:
I was thinking about this some more, and I think it would be better to just allow the user to multicall and have multiple blocks, each with different transactions.  The model may look something like:
```
[ { block_n_details, block_n_transactions }, { block_m_details, block_m_transactions }, ... ]
```
We would still require normal rules to be respected between blocks (like block numbers are incrementing, timestamp in future blocks must be higher number than previous blocks, etc.

Reply:
> or will it be some form of "time-shifted single block", where you can override the time and number, but state-processing-wise it's still the same block.... ?

This is a good point. As the implementation stands there are differences to how a sequence of blocks are executed (one being coinbase fee). As I mentioned here https://github.com/ethereum/execution-apis/pull/312#issuecomment-1295035379 I would like to proceed with "only" the single-block-multi-call variant. This would already be a big improvement for users and I would prefer not to delay that for something more complicated at the moment.

Reply:
> This would already be a big improvement for users and I would prefer not to delay that for something more complicated at the moment.

I think it is only notably more complicated if you try to do different overrides of transaction details within a single block.  My proposal is to actually have multiple blocks, each which would follow most consensus rules (like timestamps must increase, block number must increase, etc.).  I believe the complexity that Martin is referring to is specifically related to how the original proposal was designed where you have one "block" but each transaction had different block properties reflected in it.

Reply:
> I was thinking about this some more, ...
> We would still require normal rules to be respected between blocks (like block numbers are incrementing, timestamp in future blocks must be higher number than previous blocks, etc.


I've also been thinking about this some more - and I reached a different conclusion :)  In fact, kind of the opposite. I was thinking that we could make this call very much up to the caller. We would not do any form of sanity checks. If the user wants to do block `1, 500, 498, 1M, 3` in sequence, while letting `timestamp` go backwards, then fine. It's up to the caller to use this thing "correctly".

In that sense, I don't see any need to enforce "separate blocks". (To be concrete, I think that only means shipping the fees to the coinbase, so that is not a biggie really. )

I do foresee a couple of problems that maybe should be agreed with the other clients: 

1. When EVM invokes the `BLOCKHASH(number)` opcode. How should we 'resolve' blockhash when the block number is overridden. Possible semantics
   - Always return emptyhash
   - Always return `keccak256(num)`
   - Return as if it were executed on the current block, ignoring overrides. 

Currently, geth does the third option, since the `blockcontext.GetHash` function is set before any block overrides: 

```golang
	vmctx := core.NewEVMBlockContext(block.Header(), api.chainContext(ctx), nil)
	// Apply the customization rules if required.
	if config != nil {
		if err := config.StateOverrides.Apply(statedb); err != nil {
			return nil, err
		}
		config.BlockOverrides.Apply(&vmctx)
	}
```

.... I think there was one more thing I meant to write, but I've forgotten now....

Reply:
> I've also been thinking about this some more - and I reached a different conclusion :) In fact, kind of the opposite. I was thinking that we could make this call very much up to the caller. We would not do any form of sanity checks. If the user wants to do block `1, 500, 498, 1M, 3` in sequence, while letting `timestamp` go backwards, then fine. It's up to the caller to use this thing "correctly".

My concern with this strategy is that some clients (or possibly future clients) *may* be architected such that disabling basic validation checks like "block number go up" in an area that is harder to override during calls.  This is, of course, speculation on my part but it aligns with my general preference toward keeping the `multicall` as close to actual block building as possible.  I also can't think of any good use cases where having block number or time go backwards would help someone, so it feels like unnecessary leniency.

> When EVM invokes the `BLOCKHASH(number)` opcode. How should we 'resolve' blockhash when the block number is overridden. Possible semantics
> 
>     * Always return emptyhash
> 
>     * Always return `keccak256(num)`
> 
>     * Return as if it were executed on the current block, ignoring overrides.

I think there is a fourth option to include it in the potential overrides, so the caller would say "when you execute this block and `BLOCKHASH(n)` is called, return this value".  The caller could provide an array of `n` to `blockhash` values (they presumably know what set they need).  We could then fallback to one of the "reasonable defaults" that you have listed.